subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c

```diff
@@ -187,11 +213,217 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+		ok = zcbor_tstr_put_lit(zse, "len")	&&
+		     zcbor_uint64_put(zse, file_len);
+	} else {
+		/* Error, only encode error result code */
+		ok = zcbor_tstr_put_lit(zse, "rc")	&&
+		     zcbor_int32_put(zse, rc);
+	}
+
+	if (!ok) {
+		return MGMT_ERR_ENOMEM;
+	}
+
+	return 0;
+}
+#endif
+
+#if defined(CONFIG_FS_MGMT_CHECKSUM_HASH)
+/**
+ * Command handler: fs hash/checksum (read)
+ */
+static int
+fs_mgmt_file_hash_checksum(struct mgmt_ctxt *ctxt)
+{
+	char path[CONFIG_FS_MGMT_PATH_SIZE + 1];
+	char type_arr[HASH_CHECKSUM_TYPE_SIZE + 1] = CONFIG_FS_MGMT_CHECKSUM_HASH_DEFAULT;
+	char output[CONFIG_FS_MGMT_CHECKSUM_HASH_LARGEST_OUTPUT_SIZE];
+	uint64_t len = ULLONG_MAX;
+	uint64_t off = 0;
+	size_t file_len;
+	int rc;
+	zcbor_state_t *zse = ctxt->cnbe->zs;
+	zcbor_state_t *zsd = ctxt->cnbd->zs;
+	bool ok;
+	struct zcbor_string type = { 0 };
+	struct zcbor_string name = { 0 };
+	size_t decoded;
+	struct fs_file_t file;
+	const struct hash_checksum_mgmt_group *group = NULL;
+
+	struct zcbor_map_decode_key_val fs_hash_checksum_decode[] = {
+		ZCBOR_MAP_DECODE_KEY_VAL(type, zcbor_tstr_decode, &type),
+		ZCBOR_MAP_DECODE_KEY_VAL(name, zcbor_tstr_decode, &name),
+		ZCBOR_MAP_DECODE_KEY_VAL(off, zcbor_uint64_decode, &off),
+		ZCBOR_MAP_DECODE_KEY_VAL(len, zcbor_uint64_decode, &len),
+	};
+
+	ok = zcbor_map_decode_bulk(zsd, fs_hash_checksum_decode,
+		ARRAY_SIZE(fs_hash_checksum_decode), &decoded) == 0;
+
+	if (!ok || name.len == 0 || name.len > (sizeof(path) - 1) ||
+	    type.len > (sizeof(type_arr) - 1) || len == 0) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Copy strings and ensure they are null-teminated */
+	memcpy(path, name.value, name.len);
+	path[name.len] = '\0';
+
+	if (type.len != 0) {
+		memcpy(type_arr, type.value, type.len);
+		type_arr[type.len] = '\0';
+	}
+
+	/* Search for supported hash/checksum */
+	group = hash_checksum_mgmt_find_handler(type_arr);
+
+	if (group == NULL) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Check provided offset is valid for target file */
+	rc = fs_mgmt_impl_filelen(path, &file_len);
+
+	if (rc != 0) {
+		return MGMT_ERR_ENOENT;
+	}
+
+	if (file_len <= off) {
+		/* Requested offset is larger than target file size */
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Open file for reading and pass to hash/checksum generation function */
+	fs_file_t_init(&file);
+	rc = fs_open(&file, path, FS_O_READ);
+
+	if (rc != 0) {
+		return MGMT_ERR_ENOENT;
+	}
+
+	/* Seek to file's desired offset, if parameter was provided */
+	if (off != 0) {
+		rc = fs_seek(&file, off, FS_SEEK_SET);
+
+		if (rc != 0) {
+			fs_close(&file);
+			return MGMT_ERR_EINVAL;
+		}
+	}
+
+	/* Calculate hash/checksum using function */
+	file_len = 0;
+	rc = group->function(ctxt, &file, output, &file_len, len);
+
+	fs_close(&file);
+
+	/* Encode the response */
+	if (rc != 0) {
+		ok = zcbor_tstr_put_lit(zse, "rc")	&&
+		     zcbor_int32_put(zse, rc);
+	} else {
+		ok = zcbor_tstr_put_lit(zse, "type")	&&
+		     zcbor_tstr_put_term(zse, type_arr);
```

Comment:
`type` can only be used if the type was supplied, otherwise it will be empty as `type_arr` contains the default, unless you mean to do an if for which function to call and variable to use here?

Reply:
Right, I missed that. Leave it as is.